### PR TITLE
[CELEBORN-61][CELEBORN-62][FEATURE] Shuffle client support slow start, congestion avoidance and congestion control

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@boneanxs @waitinfuture

**Comment by pan3793** [comment]
what's the target version of this feature? 0.2 or 0.3?

**Comment by pan3793** [comment]
@waitinfuture I suppose currently Celeborn is during 0.2.0 release cycle, please take care which PR need to backport `branch-0.2`

**Comment by AngersZhuuuu** [comment]
> what's the target version of this feature? 0.2 or 0.3?

Should be 0.3

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 1]
```diff
@@ -665,11 +668,27 @@ public void onSuccess(ByteBuffer response) {
                               this,
                               pushState,
                               StatusCode.HARD_SPLIT));
+                } else if (reason == StatusCode.PUSH_DATA_SUCCESS_MASTER_CONGESTED.getValue()) {
+                  logger.debug(
+                      "Push data split for map {} attempt {} batch {} return master congested.",
+                      mapId,
+                      attemptId,
+                      nextBatchId);
+                  congestionControl();
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:677]: I think we also need to call callback.success here

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 2]
```diff
@@ -665,11 +668,27 @@ public void onSuccess(ByteBuffer response) {
...
+                } else if (reason == StatusCode.PUSH_DATA_SUCCESS_SLAVE_CONGESTED.getValue()) {
+                  logger.debug(
+                      "Push data split for map {} attempt {} batch {} return slave congested.",
+                      mapId,
+                      attemptId,
+                      nextBatchId);
+                  congestionControl();
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:684]: ditto

**Pull request merged**